### PR TITLE
Adds a File abstraction

### PR DESCRIPTION
**Pull request by SirVer**
This is for swapping out the underlying writing implementation, for example for writing directly to google cloud storage.

### CODE REVIEW

File: cartographer/io/pcd_writing_points_processor.cc [hunk 1]
```diff
@@ -31,69 +33,65 @@ namespace {
 // Writes the PCD header claiming 'num_points' will follow it into
 // 'output_file'.
 void WriteBinaryPcdHeader(const bool has_color, const int64 num_points,
-                          std::ofstream* const stream) {
+                          File* file) {
```
Comment (damonkohler) [inline cartographer/io/pcd_writing_points_processor.cc]: const

File: cartographer/io/file.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+ public:
+  File() {}
+  File(const File&) = delete;
+  File& operator=(const File&) = delete;
+
+  virtual ~File() {}
+
+  virtual bool Write(const char* data, size_t len) = 0;
+  virtual bool Close() = 0;
+  virtual bool SeekToStart() = 0;
```
Comment (damonkohler) [inline cartographer/io/file.h]: Given the name of the interface, I would expect a more general method (i.e. Seek). However, you're also missing Read(). Maybe choose a more descriptive name for this interface?

File: cartographer/io/file.cc [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+namespace cartographer {
+namespace io {
+
+IoStreamFile::IoStreamFile(const string& filename)
+    : out_(filename, std::ios::out | std::ios::binary) {}
+
+IoStreamFile::~IoStreamFile() {}
+
+bool IoStreamFile::Write(const char* data, size_t len) {
```
Comment (damonkohler) [inline cartographer/io/file.cc]: const len

File: cartographer/io/file.cc [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  }
+  out_.close();
+  return !out_.bad();
+}
+
+bool IoStreamFile::SeekToStart() {
+  if (out_.bad()) {
+    return false;
+  }
+  out_.flush();
```
Comment (damonkohler) [inline cartographer/io/file.cc]: It's unclear to me why you should flush before seeking.

File: cartographer/io/pcd_writing_points_processor.cc [hunk 5]
```diff
@@ -31,69 +33,65 @@ namespace {
...
+         << "POINTS " << std::setw(15) << std::setfill('0') << num_points
+         << "\n"
+         << "DATA binary\n";
+  const string out = stream.str();
+  file->Write(out.data(), out.size());
 }
 
-void WriteBinaryPcdPointCoordinate(const Eigen::Vector3f& point,
-                                   std::ofstream* const stream) {
+void WriteBinaryPcdPointCoordinate(const Eigen::Vector3f& point, File* file) {
```
Comment (damonkohler) [inline cartographer/io/pcd_writing_points_processor.cc]: const file

File: cartographer/io/pcd_writing_points_processor.cc [hunk 6]
```diff
@@ -31,69 +33,65 @@ namespace {
...
+  CHECK(file->Write(buffer, 12));
 }
 
-void WriteBinaryPcdPointColor(const Color& color, std::ostream* const stream) {
-  // Pack the color as uint32 little-endian
-  stream->put(color[2]);
-  stream->put(color[1]);
-  stream->put(color[0]);
-  stream->put(0);
+void WriteBinaryPcdPointColor(const Color& color, File* file) {
```
Comment (damonkohler) [inline cartographer/io/pcd_writing_points_processor.cc]: const file

File: cartographer/io/ply_writing_points_processor.cc [hunk 7]
```diff
@@ -31,63 +33,55 @@ namespace {
 // Writes the PLY header claiming 'num_points' will follow it into
 // 'output_file'.
 void WriteBinaryPlyHeader(const bool has_color, const int64 num_points,
-                          std::ofstream* const stream) {
+                          File* file) {
```
Comment (damonkohler) [inline cartographer/io/ply_writing_points_processor.cc]: const

File: cartographer/io/ply_writing_points_processor.cc [hunk 8]
```diff
@@ -31,63 +33,55 @@ namespace {
...
+         << "property float y\n"
+         << "property float z\n"
+         << color_header << "end_header\n";
+  const string out = stream.str();
+  CHECK(file->Write(out.data(), out.size()));
 }
 
-void WriteBinaryPlyPointCoordinate(const Eigen::Vector3f& point,
-                                   std::ofstream* const stream) {
+void WriteBinaryPlyPointCoordinate(const Eigen::Vector3f& point, File* file) {
```
Comment (damonkohler) [inline cartographer/io/ply_writing_points_processor.cc]: const file (there are many of these, will stop commenting now)

File: cartographer/io/points_processor_pipeline_builder.cc [hunk 9]
```diff
@@ -42,32 +42,51 @@ void RegisterPlainPointsProcessor(
...
       });
 }
 
 void PointsProcessorPipelineBuilder::Register(const std::string& name,
                                               FactoryFunction factory) {
-  CHECK(factories_.count(name) == 0) << "A points processor with named '"
-                                     << name
-                                     << "' has already been registered.";
+  CHECK(factories_.count(name) == 0)
+      << "A points processor with named '" << name
```
Comment (damonkohler) [inline cartographer/io/points_processor_pipeline_builder.cc]: with name or named

File: cartographer/io/xray_points_processor.cc [hunk 10]
```diff
@@ -48,8 +48,17 @@ void TakeLogarithm(Eigen::MatrixXf* mat) {
   }
 }
 
+cairo_status_t CairoWriteCallback(void* closure, const unsigned char* data,
```
Comment (damonkohler) [inline cartographer/io/xray_points_processor.cc]: consts

File: cartographer/io/xray_points_processor.cc [hunk 11]
```diff
@@ -150,7 +162,8 @@ XRayPointsProcessor::XRayPointsProcessor(
 
 std::unique_ptr<XRayPointsProcessor> XRayPointsProcessor::FromDictionary(
     const mapping::proto::Trajectory& trajectory,
-    common::LuaParameterDictionary* dictionary, PointsProcessor* next) {
+    const FileFactory& file_factory, common::LuaParameterDictionary* dictionary,
```
Comment (damonkohler) [inline cartographer/io/xray_points_processor.cc]: consts

File: cartographer/io/file.cc (hunk 4 above)
Comment (SirVer) [inline cartographer/io/file.cc]: To write out remaining bytes in the queue before seeking to the beginning of it? This file is write only.

File: cartographer/io/file.h (hunk 2 above)
Comment (SirVer) [inline cartographer/io/file.h]: I mean, it is strictly write at this point in time. What would you suggest? FileWrite?

Comment (damonkohler) [inline cartographer/io/file.h]: FileWriter?

File: cartographer/io/file.cc (hunk 4 above)
Comment (damonkohler) [inline cartographer/io/file.cc]: Please replace this with a more meaningful function. Something like WriteHeader()?

**Comment by SirVer** [comment]
Made the changes discussed offline.

File: cartographer/io/file_writer.h [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+class FileWriter {
+ public:
+  FileWriter() {}
+  FileWriter(const FileWriter&) = delete;
+  FileWriter& operator=(const FileWriter&) = delete;
+
+  virtual ~FileWriter() {}
+
+  // Write 'data' to the beginning of the file. This is required to overwrite
+  // fix sized headers which contain the number of points once we actually know
```
Comment (damonkohler) [inline cartographer/io/file_writer.h]: fixed size

File: cartographer/io/pcd_writing_points_processor.cc [hunk 13]
```diff
@@ -2,7 +2,7 @@
  * Copyright 2016 The Cartographer Authors
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
- * you may not use this file except in compliance with the License.
+ * you may not use this file_writer except in compliance with the License.
```
Comment (damonkohler) [inline cartographer/io/pcd_writing_points_processor.cc]: Double check your global search/replace.

File: cartographer/io/xray_points_processor.cc [hunk 14]
```diff
@@ -76,11 +86,13 @@ void WritePng(const string& filename, const Eigen::MatrixXf& mat) {
...
   CHECK_EQ(cairo_surface_status(surface.get()), CAIRO_STATUS_SUCCESS);
-  CHECK_EQ(cairo_surface_write_to_png(surface.get(), filename.c_str()),
+  CHECK_EQ(cairo_surface_write_to_png_stream(surface.get(), &CairoWriteCallback,
+                                             file_writer),
            CAIRO_STATUS_SUCCESS);
+  CHECK(file_writer->Close());
 }
 
-void WriteVoxels(const string& filename, const Voxels& voxels) {
+void WriteVoxels(const Voxels& voxels, FileWriter* file_writer) {
```
Comment (damonkohler) [inline cartographer/io/xray_points_processor.cc]: const

File: cartographer/io/xray_points_processor.cc [hunk 15]
```diff
@@ -138,8 +150,9 @@ void Insert(const PointsBatch& batch, const transform::Rigid3f& transform,
 XRayPointsProcessor::XRayPointsProcessor(
     const double voxel_size, const transform::Rigid3f& transform,
     const std::vector<mapping::Floor>& floors, const string& output_filename,
-    PointsProcessor* next)
+    const FileWriterFactory& file_writer_factory, PointsProcessor* next)
```
Comment (damonkohler) [inline cartographer/io/xray_points_processor.cc]: const next

File: cartographer/io/xyz_writing_points_processor.cc [hunk 16]
```diff
@@ -10,29 +10,31 @@ namespace io {
 
 namespace {
 
-void WriteXyzPoint(const Eigen::Vector3f& point, std::ofstream* output) {
-  (*output) << point.x() << " " << point.y() << " " << point.z() << "\n";
+void WriteXyzPoint(const Eigen::Vector3f& point, FileWriter* file_writer) {
```
Comment (damonkohler) [inline cartographer/io/xyz_writing_points_processor.cc]: const

File: cartographer/io/xyz_writing_points_processor.cc [hunk 17]
```diff
@@ -10,29 +10,31 @@ namespace io {
...
 
 }  // namespace
 
-XyzWriterPointsProcessor::XyzWriterPointsProcessor(const string& filename,
-                                                   PointsProcessor* next)
-    : next_(next), file_(filename, std::ios_base::out | std::ios_base::binary) {
-  file_ << std::setprecision(6);
-}
+XyzWriterPointsProcessor::XyzWriterPointsProcessor(
+    std::unique_ptr<FileWriter> file_writer, PointsProcessor* next)
```
Comment (damonkohler) [inline cartographer/io/xyz_writing_points_processor.cc]: const

**Pull request merged**